File: packages/swift/Sources/CatalystSwift/Models/EpicleseEndpoint.swift
// Licensed under the MIT License
//
// Copyright (c) 2025 Natsune Mochizuki

public enum EpicleseEndpoint: Endpoint {
  case getAuthors(q: String?, platform: String?)
  case createAuthor(data: EpicleseCreateAuthorRequest)
  case getPlatforms
  case getPlatform(id: String)
  case getStatusMetadata(statusId: String)
  case createStatusMetadata(statusId: String, data: [EpicleseCreateStatusMetadataTagRequest])
  case getWorlds(q: String?, platform: String?, offset: Int?)
  case createWorld(data: EpicleseCreateWorldRequest)
  case resolveWorld(platform: String, name: String)

  public var path: String {
    switch self {
    case .getAuthors, .createAuthor:
      return "/epiclese/v1/authors"

    case .getPlatforms:
      return "/epiclese/v1/platforms"

    case .getPlatform(let id):
      return "/epiclese/v1/platforms/\(id)"

    case .getStatusMetadata(let statusId), .createStatusMetadata(let statusId, _):
      return "/epiclese/v1/tag/by/status/\(statusId)"

    case .getWorlds, .createWorld:
      return "/epiclese/v1/worlds"

    case .resolveWorld:
      return "/epiclese/v1/worlds/resolve"
    }
  }

  public var method: HTTPMethod {
    switch self {
    case .getAuthors, .getPlatforms, .getPlatform, .getStatusMetadata, .getWorlds, .resolveWorld:
      return .get

    case .createAuthor, .createStatusMetadata, .createWorld:
      return .post
    }
  }

  public var headers: [String: String]? {
    return nil
  }

  public var queryParameters: [String: String]? {
    switch self {
    case .getAuthors(let q, let platform):
      var params: [String: String] = [:]
      if let q { params["q"] = q }
      if let platform { params["platform"] = platform }
      return params.isEmpty ? nil : params

    case .getWorlds(let q, let platform, let offset):
      var params: [String: String] = [:]
      if let q { params["q"] = q }
      if let platform { params["platform"] = platform }
      if let offset { params["offset"] = String(offset) }
      return params.isEmpty ? nil : params

    case .resolveWorld(let platform, let name):
      return ["platform": platform, "name": name]

    default:
      return nil
    }
  }

  public var body: Encodable? {
    switch self {
    case .createAuthor(let data):
      return data

    case .createStatusMetadata(_, let data):
      return data

    case .createWorld(let data):
      return data

    default:
      return nil
    }
  }
}

File: packages/nodejs/src/endpoints/epicleseEndpoint.ts
import type { Endpoint } from "./endpoint.js";
import type {
  EpicleseCreateAuthorRequest,
  EpicleseCreateWorldRequest,
  EpicleseCreateStatusMetadataTagRequest,
} from "../types/epiclese.js";

export const EpicleseEndpoint = {
  getAuthors(q?: string, platform?: string): Endpoint {
    const params: Record<string, string> = {};
    if (q != null) params["q"] = q;
    if (platform != null) params["platform"] = platform;
    return {
      path: "/epiclese/v1/authors",
      method: "GET",
      queryParameters: Object.keys(params).length > 0 ? params : undefined,
    };
  },

  createAuthor(data: EpicleseCreateAuthorRequest): Endpoint {
    return { path: "/epiclese/v1/authors", method: "POST", body: data };
  },

  getPlatforms(): Endpoint {
    return { path: "/epiclese/v1/platforms", method: "GET" };
  },

  getPlatform(id: string): Endpoint {
    return { path: `/epiclese/v1/platforms/${id}`, method: "GET" };
  },

  getStatusMetadata(statusId: string): Endpoint {
    return { path: `/epiclese/v1/tag/by/status/${statusId}`, method: "GET" };
  },

  createStatusMetadata(
    statusId: string,
    data: EpicleseCreateStatusMetadataTagRequest[],
  ): Endpoint {
    return {
      path: `/epiclese/v1/tag/by/status/${statusId}`,
      method: "POST",
      body: data,
    };
  },

  getWorlds(q?: string, platform?: string, offset?: number): Endpoint {
    const params: Record<string, string> = {};
    if (q != null) params["q"] = q;
    if (platform != null) params["platform"] = platform;
    if (offset != null) params["offset"] = String(offset);
    return {
      path: "/epiclese/v1/worlds",
      method: "GET",
      queryParameters: Object.keys(params).length > 0 ? params : undefined,
    };
  },

  createWorld(data: EpicleseCreateWorldRequest): Endpoint {
    return { path: "/epiclese/v1/worlds", method: "POST", body: data };
  },

  resolveWorld(platform: string, name: string): Endpoint {
    return {
      path: "/epiclese/v1/worlds/resolve",
      method: "GET",
      queryParameters: { platform, name },
    };
  },
} as const;
